### PR TITLE
[fix][broker] Fix direct memory leak by delayed index OutOfDirectMemory

### PR DESCRIPTION
**Pull request by mattisonchao**
### Motivation

[There](https://github.com/apache/pulsar/blob/57fbee446aebb5c9f4db3bed559cb7bef3749731/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java#L205C16-L205C16) is a memory leak if the delayed tracker throws an `OutOfDirectMemory` exception. The entries will never release. 

Plus, with the OOM exception thrown in the thread, the broker will not shut down.

```

Jul 16, 2023 @ 18:35:58.575 | 2023-07-16T10:35:58,574+0000 [broker-topic-workers-OrderedExecutor-0-0] ERROR org.apache.bookkeeper.common.util.SingleThreadExecutor - Error while running task: failed to allocate 16777216 byte(s) of direct memory (used: 847249415, max: 858993459)
  | Jul 16, 2023 @ 18:35:58.575 | at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[io.netty-netty-common-4.1.93.Final.jar:4.1.93.Final]
  | Jul 16, 2023 @ 18:35:58.575 | at org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers.trySendMessagesToConsumers(PersistentDispatcherMultipleConsumers.java:717) ~[io.streamnative-pulsar-broker-3.0.0.4.jar:3.0.0.4]
  | Jul 16, 2023 @ 18:35:58.575 | at org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers.trackDelayedDelivery(PersistentDispatcherMultipleConsumers.java:1068) ~[io.streamnative-pulsar-broker-3.0.0.4.jar:3.0.0.4]
  | Jul 16, 2023 @ 18:35:58.575 | at io.netty.buffer.PoolArena.allocate(PoolArena.java:144) ~[io.netty-netty-buffer-4.1.93.Final.jar:4.1.93.Final]
  | Jul 16, 2023 @ 18:35:58.575 | at io.netty.util.internal.PlatformDependent.allocateDirectNoCleaner(PlatformDependent.java:774) ~[io.netty-netty-common-4.1.93.Final.jar:4.1.93.Final]
  | Jul 16, 2023 @ 18:35:58.575 | at org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers.sendMessagesToConsumers(PersistentDispatcherMultipleConsumers.java:622) ~[io.streamnative-pulsar-broker-3.0.0.4.jar:3.0.0.4]
  | Jul 16, 2023 @ 18:35:58.575 | at org.apache.pulsar.broker.delayed.InMemoryDelayedDeliveryTracker.addMessage(InMemoryDelayedDeliveryTracker.java:81) ~[io.streamnative-pulsar-broker-3.0.0.4.jar:3.0.0.4]
  | Jul 16, 2023 @ 18:35:58.575 | at io.netty.buffer.PoolArena.allocate(PoolArena.java:129) ~[io.netty-netty-buffer-4.1.93.Final.jar:4.1.93.Final]
  | Jul 16, 2023 @ 18:35:58.575 | at io.netty.buffer.PoolArena$DirectArena.newUnpooledChunk(PoolArena.java:690) ~[io.netty-netty-buffer-4.1.93.Final.jar:4.1.93.Final]
  | Jul 16, 2023 @ 18:35:58.575 | io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 16777216 byte(s) of direct memory (used: 847249415, max: 858993459)
  | Jul 16, 2023 @ 18:35:58.575 | at org.apache.bookkeeper.common.util.SingleThreadExecutor.safeRunTask(SingleThreadExecutor.java:137) ~[org.apache.bookkeeper-bookkeeper-common-4.16.2.jar:4.16.2] 
  | Jul 16, 2023 @ 18:35:58.575 | at org.apache.pulsar.common.util.collections.TripleLongPriorityQueue.add(TripleLongPriorityQueue.java:94) ~[io.streamnative-pulsar-common-3.0.0.4.jar:3.0.0.4]
  | Jul 16, 2023 @ 18:35:58.575 | at io.netty.buffer.PoolArena.allocateHuge(PoolArena.java:226) ~[io.netty-netty-buffer-4.1.93.Final.jar:4.1.93.Final]
  | Jul 16, 2023 @ 18:35:58.575 | at org.apache.bookkeeper.common.util.SingleThreadExecutor.run(SingleThreadExecutor.java:113) ~[org.apache.bookkeeper-bookkeeper-common-4.16.2.jar:4.16.2]
```

### Modifications

- Using `PulsarByteBufAllocator` instead `PooledByteBufAllocator` to have better OOM control.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

**Review by coderzc — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java [hunk 1]
```diff
@@ -1065,7 +1066,17 @@ public boolean trackDelayedDelivery(long ledgerId, long entryId, MessageMetadata
             delayedDeliveryTracker.get().resetTickTime(topic.getDelayedDeliveryTickTimeMillis());
 
             long deliverAtTime = msgMetadata.hasDeliverAtTime() ? msgMetadata.getDeliverAtTime() : -1L;
-            return delayedDeliveryTracker.get().addMessage(ledgerId, entryId, deliverAtTime);
+            try {
+                return delayedDeliveryTracker.get().addMessage(ledgerId, entryId, deliverAtTime);
+            } catch (OutOfDirectMemoryError ex) {
```
Comment (poorbarcode) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:1071]: In general, the Error type is not recommended to catch, we should use these two ways to solve the problem
- Set the maximum memory usage of the delay message. If the delay message usage exceeds the maximum memory usage, stuck new read
- we should kill the process if the error occurs

Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:1071]: The current implementation is the first approach, but we don't introduce a fixed value.  Instead, we use a dynamic value to do it.

> Set the maximum memory usage of the delay message.
This is hard to set a reasonable value.

> stuck new read
when we recover?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java [hunk 2]
```diff
@@ -1065,7 +1066,17 @@ public boolean trackDelayedDelivery(long ledgerId, long entryId, MessageMetadata
...
+                log.error("Out of direct memory while trying add message index to delayed tracker."
+                        + " fallback to redelivery messages(heap) and waiting for enough direct memory space."
+                        + " error info: {}", ex.getMessage());
+                // fall back to redelivery queue(heap) waiting for enough direct memory.
+                redeliveryMessages.add(ledgerId, entryId);
```
Comment (poorbarcode) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:1076]: This makes the delayed messages never delay, right?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java (hunk 1 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:1071]: Just like the implementation of JVM, if could not allocate more memory, the progress will be killed by OOM. It did not do `yield` and try another once.

>> Set the maximum memory usage of the delay message. If the delay message usage exceeds the maximum memory usage, stuck new read
> This is an ideal approach. what is the good time value to recover? 

In this patch, you mv this message into `replay queue`, and the next reading will get it again, loop. The result is actually the same as doing a stuck new read.

**Review by poorbarcode — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

**Review by Technoboy- — APPROVED** [review]

**Pull request merged**